lib/iris/_shapefiles.py

```diff
@@ -0,0 +1,223 @@
+# the Met Office by Chris Kent, Emilie Vanvyve, David Bentley, Joana Mendes
+# many thanks to them. Converted to iris by Alex Chamberlain-Clay s
+
+
+import warnings
+
+import numpy as np
+import shapely
+import shapely.errors
+import shapely.geometry as sgeom
+import shapely.ops
+import shapely.prepared as prepped
+
+from iris.exceptions import IrisDefaultingWarning
+
+
+def create_shapefile_mask(
+    geometry, cube, minimum_weight=0.0, shape_coord_system=None
+):
+    """Get the mask array of the intersection between the given shapely geometry and
+    cube.
+
+    Arguments:
+        geometry        : A :class:`shapely.Geometry` object
+
+        cube            : A :class:`iris.cube.Cube`
+                            with x and y coordinates
+        minimum_weight  : A float between 0 and 1 determining what % of a cell
+                            a shape must cover for the cell to remain unmasked.
+                            eg: 0.1 means that at least 10% of the shape overlaps the cell
+                            to be unmasked
+
+    Returns:
+        A numpy array of the shape of the x & y coordinates of the cube, with points to mask equal to True
+    """
+
+    from iris.cube import Cube, CubeList
+
+    try:
+        msg = TypeError("Geometry is not a valid Shapely object")
+        if geometry.is_valid is False:
+            raise msg
+    except Exception:
+        raise msg
+    if not isinstance(cube, Cube):
+        if isinstance(cube, CubeList):
+            msg = "Received CubeList object rather than Cube - \
+            to mask a CubeList iterate over each Cube"
+            raise TypeError(msg)
+        else:
+            msg = "Received non-Cube object where a Cube is expected"
+            raise TypeError(msg)
+    if not minimum_weight or minimum_weight == 0.0:
+        weights = False
+    elif isinstance(
+        geometry,
+        (
+            sgeom.Point,
+            sgeom.LineString,
+            sgeom.LinearRing,
+            sgeom.MultiPoint,
+            sgeom.MultiLineString,
+        ),
+    ):
+        weights = False
+        warnings.warn(
+            """Shape is of invalid type for minimum weight masking,
+            must use a Polygon rather than Line shape.\n
+              Masking based off intersection instead. """,
+            IrisDefaultingWarning,
+        )
+    else:
+        weights = True
+
+    # prepare shape
+    trans_geo = _transform_coord_system(geometry, cube, shape_coord_system)
+    prepped_trans_geo = prepped.prep(trans_geo)
+
+    # prepare 2D cube
+    y_name, x_name = _cube_primary_xy_coord_names(cube)
+    cube_2d = cube.slices([y_name, x_name]).next()
+    _cube_xy_guessbounds(cube_2d)
+    xmod = cube.coord(x_name).units.modulus
+    ymod = cube.coord(y_name).units.modulus
+    mask_template = np.zeros(cube_2d.shape, dtype=bool)
+
+    # perform the masking
+    for count, idx in enumerate(np.ndindex(cube_2d.shape)):
```

Comment:
Did you consider using an equivalent of ASCEND's `Shape.constrain_cube()` upstream, to minimise the scope of the cell-by-cell checking?

Reply:
Had this thought - it requires too much code to implement for the marginal performance improvement, effort better spent on using dask for parralisation which is more universal, rather than constrain_cube() which relies on the shape being regular to give any improvement.

Reply:
I've not really been following closely, so perhaps I'm missing something, but I'm not sure I understand why you wouldn't do this? If you constrain to the **bounding box** of the shape, you have much less of the cube to check - everything outside the bounding box you can automatically mask, so you only need to loop through the constrained cube. You just then need to appropriately index your `mask_template`, but it shouldn't be that much code.